Rework OrderHandlers calculations to use Prestashop Number

zuk3975 opened the pull request:
<!--
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
 -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.7.7.x
| Description?  | See below
| Type?         | refacto
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | part of #17282 and #18020
| How to test?  | See below

## Description

### Initial scope

- fix `@todo` in **AddCartRuleToOrderHandler** ('move to private method' in couple places)

### Scope expansion

- refactors **AddCartRuleToOrderHandler**, **AddProductToOrderHandler**, **DeleteProductFromOrderHandler**, and **UpdateProductInOrderHandler** handlers into private methods
- implement Number calculations instead of floats for handlers **AddCartRuleToOrderHandler**, **AddProductToOrderHandler**, **DeleteProductFromOrderHandler**, and **UpdateProductInOrderHandler**
- added some behat tests
- removed the ability to update invoices when adding discount to order _as requested by PM_
- fixed bug where "free shipping" checkbox wasn't displayed when adding product with new invoice

## How to test ?

Use `sell/orders/orders/[orderId]/view` to reach migrated order view page. Make sure `Add discount` is working as expected.
**Tip:** to add product with new invoice the order status must be "payment accepted" (works with some others as well).
Be sure to check the product total prices (shipping, total, taxes, discounts- _this one also increments when product is added with free shipping option_).
:warning:  **WARNING**
Pr included a heavy rebase associated with price calculations from #17877 and #17476. So its important to test according to those PR's as well.

## Known issues
* when adding product with free shipping, the discounts are increased, but not shipping price, which leads to wrong total numbers. (the overall total for the order is correct though). _(this one is hard to fix and is delayed to at least 1.7.8)_
* After adding product with new invoice and free shipping, the page needs to be refreshed to see the new specific price rule in list. _(this should be solved in another PR)_

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/17348)
<!-- Reviewable:end -->

zuk3975 committed:
refactor handler to private methods

matks commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,82 +50,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
-            $orderInvoice = new OrderInvoice($command->getOrderInvoiceId()->getValue());
-            if (!Validate::isLoadedObject($orderInvoice)) {
-                throw new OrderException('Can\'t load Order Invoice object');
+        $discountValue = (float) str_replace(',', '.', $command->getDiscountValue());
```
😱

zuk3975 replied:
Sorry guys, I have no knowledge about whats happening there deep, I just tried to at least separate to private methods, didn't wan't to break the logic.

jolelievre replied:
It was used in the legacy code to reformat the received string, but since the command has a strong typed float it's useless

zuk3975 replied:
It is typed as null|Number . If I use Number I need to rework all calculations, thats risky and alot of code. You think its worth it ?

matks commented:
That would be very cool to have some behat tests to help us validate this complex and critical code. For a later PR maybe ?

matks commented:
Closing and reopening, it seems Travis is checking against `develop` and not against `1.7.7.x`

PierreRambaud commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,82 +50,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
+            $cartRuleObj = $this->createCartRule($command, $order, $cartRule, $discountValue);
+
+            if (false === $cartRuleObj->add()) {
+                break;
```
should we really need to break?

PierreRambaud replied:
what if we add nine rules, but the ten one is broken, we will have an exception, what about the nine rules?

zuk3975 replied:
Fixed. (not sure why its not updating here):
But it looks like it previously was:
```
            if ($result = $cartRuleObj->add()) {
                $cartRule['id'] = $cartRuleObj->id;
            } else {
                break;
            }
```

zuk3975 replied:
As I understand the command adds only one cart rule.
I think its all wrong with the naming, It should be singular.
buildCartRulesByType is only checking which type of cartRule should be built.

Ill make sure its that, and update it..

PierreRambaud replied:
It's disturbing `AddCartRuleToOrderHandler` should be `AddCartRulesToOrderHandler` or the `buildCartRulesByType` should be singular and receive only one cart rule (and the foreach became useless) :sweat_smile:

zuk3975 replied:
yep, trying to figure out whats happening, its confusing cuz its mostly legacy code pasted to handler :smile:

It should create only one cartRule, but multiple invoices has something to do with multiple cartRules, so I don't really know how should it behave and how the order gets multiple invoices in first place.

jolelievre replied:
It happens when you add a product to an order which already has an invoice You can chosse to add the new product to a new invoice..

PierreRambaud commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,82 +50,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
+            $cartRule['id'] = $cartRuleObj->id;
+        }
+
+        if ($result) {
```
$result is always true here :thinking:

zuk3975 replied:
Fixed above:
```
            if ($result = $cartRuleObj->add()) {
                $cartRule['id'] = $cartRuleObj->id;
            } else {
                break;
            }
```
Original stmt from previous pr.

PierreRambaud commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,82 +50,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
+     */
+    private function buildPercentTypeCartRules(array &$cartRules, Order $order, float $discountValue, ?OrderInvoice $orderInvoice): void
+    {
+        if ($discountValue < 100) {
```
why 100, must be a constant

zuk3975 committed:
add const MAX_PERCENTAGE

zuk3975 commented:
addressed comments.

PierreRambaud commented:
Thanks @zuk3975. Because it's a migration work, I'll not annoy you anymore :sweat_smile:

zuk3975 commented:
> Thanks @zuk3975. Because it's a migration work, I'll not annoy you anymore

hmm sorry, you should :smile: I've noticed It still needs some changes. see https://github.com/PrestaShop/PrestaShop/pull/17348#discussion_r372266207

zuk3975 commented:
@matks, the code is crappy inside the adapter (it was like pasted straight from legacy before I touched it), (most confusing part is that command says its adding single CartRule, but it does create an array of cart rules based on invoices inside..), but I don't know how it suppose to work, so I think I will not try to improve it anymore, as I might break smth, and its is not what we want in featurefreeze.
However, I could rework this if i have clear specifications about discount creation (and use cases with single/multiple invoices)

matks commented:
> @matks, the code is crappy inside the adapter (it was like pasted straight from legacy before I touched it - most confusing part is that command says its adding single CartRule, but it does create an array of cart rules based on invoices inside..), but I don't know how it suppose to work, so I think I will not try to improve it anymore, as I might break smth, and its is not what we want in featurefreeze.
> However, I could rework this if i have clear specifications about discount creation (and use cases with single/multiple invoices)

Let's do our best 😢 

If PM writes such specs, indeed we'll be able to improve this

jolelievre commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,82 +51,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
     public function handle(AddCartRuleToOrderCommand $command): void
     {
         $order = $this->getOrderObject($command->getOrderId());
+        $orderInvoice = $this->createSingleOrderInvoice($order, $command);
```
It would be clearer if you managed the `single order invoice` case here, if `createSingleOrderInvoice` return null it means you have to apply the discount to all its invoices. So you could do:
```php
$orderInvoice = $this->createSingleOrderInvoice($order, $command);
if (null !== $orderInvoice) {
    $orderInvoices = [$orderInvoice];
} else {
    $orderInvoices = $order-> getInvoicesCollection()->getAll();
}
```

This way you always have an array of OrderInvoice objects, and you can simplify most of your methods, you don't need to transfer the Order and a potentially null OrderInvoice. The method `buildCartRulesByType` (and its subsequent methods) can be simplified into `buildCartRulesByType($orderInvoices, $discountValue, $cartRuleType)`

jolelievre commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,82 +51,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
-                } else {
-                    throw new OrderException('Percentage discount value cannot be higher than 100%.');
-                }
+                $this->buildPercentTypeCartRules($cartRules, $order, $discountValue, $orderInvoice);
```
I don't think you need to instanciate `$cartRules` beforehand and pass it around in each method Since you only deal with one cart rule type, only one method will be called, so each methods `buildPercentTypeCartRules`,  `buildAmountTypeCartRules` and `buildPercentTypeCartRules` can create its own internal array and return it.

This way (if you already made the simplification I suggested earlier with the array of invoices) you can simply do:

```php
switch ($cartRuleType) {
    case OrderDiscountType::DISCOUNT_PERCENT:
        return $this->buildPercentTypeCartRules($orderInvoices, $discountValue);
        break;
    ....
}
```

jolelievre commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,82 +51,137 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
+    private function buildPercentTypeCartRules(array &$cartRules, Order $order, float $discountValue, ?OrderInvoice $orderInvoice): void
+    {
+        if ($discountValue < PercentageDiscount::MAX_PERCENTAGE) {
+            if (isset($orderInvoice)) {
```
If you only deal with an array of OrderInvoice this method can be simplified (the other two as well) you wouldn't need to deal with three different cases each time only:
```php
if (count($orderInvoices) > 0) {
 ...
} else {
 ...
}
```

jolelievre commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -134,150 +191,196 @@ public function handle(AddCartRuleToOrderCommand $command): void
+     * @return CartRule
+     */
+    private function createCartRule(
+        AddCartRuleToOrderCommand $command,
```
You can simply use the cart rule type as a string instead of passing around the command all the time

jolelievre commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -134,150 +191,196 @@ public function handle(AddCartRuleToOrderCommand $command): void
-     * @param OrderInvoice $orderInvoice
-     * @param float $valueTaxIncl
-     * @param float $valueTaxExcl
+     * @param AddCartRuleToOrderCommand $command
```
You can simplify by using:

```php
/**
 * @param string $cartRuleName
 * @param int $orderId
*/
```

jolelievre reviewed (changes requested):
This is not an easy code to say the least ^^
I made a few suggestion that will allow to reduce the size of the code and will make it more readable I think, let me know what you think of it

matks commented:
As said by @zuk3975, 
> but I don't know how it suppose to work, so I think I will not try to improve it anymore, as I might break something, and its is not what we want in feature freeze

This is unreadable and untested legacy code that we want only to move. I strongly suggest not to touch it.

zuk3975 commented:
@jolelievre addressed comments, I think its enough for this one, a bit clearer now :smile: need QA to retest (although i tried to some usecases and it seem to work)

zuk3975 commented:
I will add some behats for this in same pr

zuk3975 committed:
behat background and scenario with amount discount

zuk3975 committed:
behat percent type discount no invoice

zuk3975 committed:
Add behat cases for single/multiple invoices. Failing test - need to clarify exected behavior

zuk3975 commented:
@matks, @MatShir I've added some behat but noticed kinda strange behavior, and wanted to clarify.
So, when adding a product to order with new invoice there is a checkbox for "free shipping". 
1.Should the shipping price be applied to the product alone or override the whole order shipping and recalculate?
Currently it is applied only to added product.
2. If I select free shipping, shouldn't it be reflected in discount total and shipping total of the whole order?
Currently it is only reflected in Total price of order, but doesn't add up to total shipping.

Need your answers to fix the behavior (if needed) and finish the pr, thanks.

zuk3975 commented:
> @matks, @MatShir I've added some behat but noticed kinda strange behavior, and wanted to clarify.
> So, when adding a product to order with new invoice there is a checkbox for "free shipping".
> 1.Should the shipping price be applied to the product alone or override the whole order shipping and recalculate?
> Currently it is applied only to added product.
> 2. If I select free shipping, shouldn't it be reflected in discount total and shipping total of the whole order?
> Currently it is only reflected in Total price of order, but doesn't add up to total shipping.
> 
> Need your answers to fix the behavior (if needed) and finish the pr, thanks.

Ok so, we had sesion with @MatShir, we found out that the calculations of shipping, discounts and sometimes a total price is wrong. Going to figure it out and fix it in the same PR.

zuk3975 committed:
fix order totals when adding product to order

zuk3975 committed:
rework Adding product and cart rule to order. Update tests

zuk3975 committed:
Add more specific behat scenarios

 apply cs fixer on tests

zuk3975 commented:
@jolelievre, @matks ready for review again

Progi1984 commented:
@zuk3975 Please fix Travis CI tests : Behat test from `Features/Scenario/Order/add_cart_rule_to_order.feature` are all broken.

zuk3975 commented:
> @zuk3975 Please fix Travis CI tests : Behat test from `Features/Scenario/Order/add_cart_rule_to_order.feature` are all broken.

Yeah im on it, it was something wrong with local test database fixtures.

zuk3975 committed:
update tests with correct product price

zuk3975 commented:
@jolelievre , @Progi1984 , @matks, @PierreRambaud waiting for approve

rokaszygmantas commented on tests/Integration/Behaviour/Features/Scenario/Order/add_cart_rule_to_order.feature:
```diff
@@ -0,0 +1,325 @@
+    And the module "dummy_payment" is installed
+    And I am logged in as "test@prestashop.com" employee
+    And I add 2 products "Mug The best is yet to come" to the cart "dummy_cart"
+    And I add order "bo_order1" with the following details:
```
```suggestion
    And I place order "bo_order1" with the following details:
```

zuk3975 replied:
I doubt its in pr scope, because those were implemented earlier and just reused here.

jolelievre replied:
This one makes sense, but indeed not in this PR scope I think

rokaszygmantas commented on tests/Integration/Behaviour/Features/Scenario/Order/add_cart_rule_to_order.feature:
```diff
@@ -0,0 +1,325 @@
+    And cart "dummy_cart" delivery and invoice address country for customer "testCustomer" is "US"
+    And the module "dummy_payment" is installed
+    And I am logged in as "test@prestashop.com" employee
+    And I add 2 products "Mug The best is yet to come" to the cart "dummy_cart"
```
```suggestion
    And I add 2 quantity of product "Mug The best is yet to come" to the cart "dummy_cart"
```

zuk3975 replied:
I doubt its in pr scope, because those were implemented earlier and just reused here.

jolelievre replied:
Besides it's not easier to read

jolelievre commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,235 +52,296 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
+            if ($result = $cartRuleObj->add()) {
+                $reducedValues['cart_rule_id'] = $cartRuleObj->id;
+            } else {
+                break;
```
This `break` happens when a cart rule could not be added, then `$result == false` and an exception is thrown L95
To make the code clearer could you please throw the exception here instead of the `break` ?

jolelievre commented on src/Adapter/Order/CommandHandler/AddCartRuleToOrderHandler.php:
```diff
@@ -49,235 +52,296 @@ final class AddCartRuleToOrderHandler extends AbstractOrderHandler implements Ad
+            $orderInvoices = $order->getInvoicesCollection()->getResults();
+        }
+
+        $discountValue = (float) $command->getDiscountValue()->toPrecision(CommonAbstractType::PRESTASHOP_DECIMALS);
```
This should be related to the currency, using the `ComputingPrecision` class, but this can be improved later 😉

jolelievre commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -498,27 +414,91 @@ private function createNewInvoice(Order $order, Cart $cart, $isFreeShipping)
+        $order->total_shipping_tax_excl += (float) abs($cart->getOrderTotal(false, Cart::ONLY_SHIPPING));
+    }
+
+    private function createFreeShippingCartRule(Cart $cart, Order $order): CartRule
```
```suggestion
    private function addFreeShippingCartRule(Cart $cart, Order $order): CartRule
```

jolelievre reviewed (commented):
A few feedback but nothing blocking

Robin-Fischer-PS commented:
Hi @zuk3975 :)

I have two bugs : 
- When adding a Free Shipping, an Exception is displayed
![pr17348 free ship](https://user-images.githubusercontent.com/55688534/74854396-92410080-533f-11ea-84e2-e4fbd05418e6.png)
- When choosing the "Apply on all invoices", the discount is applied only once, but displayed once for each invoice. It should be applied as many times as there is invoices
![pr17348 all invoice voucher](https://user-images.githubusercontent.com/55688534/74856193-04b2e000-5342-11ea-9235-b37240989bae.png)

Thanks !

zuk3975 committed:
fix wrong docblock

zuk3975 commented:
> When choosing the "Apply on all invoices", the discount is applied only once, but displayed once for each invoice. It should be applied as many times as there is invoices

Sorry this can't be right ( I know it worked like this in legacy, but it doesn't sound right).
@colinegin, @sam-pires could you please take a look? (_We have discussed it abit in chat, but it would be great to see some recap here_)

colinegin commented:
So just a feedback about invoice edition : this is normal that we allow it, as it is not illegal to edit an invoice (for instance delivery address) even though the payment has been accepted. It should not be done when the order is too old, but there is no precise information about what details can be edited and until when it can be modified. So the only solution we have is to let the user edit what he wants, it's his responsibility to be compliant with the law.

About the "apply to all invoices" button, I agree with @jolelievre, I recommend to drop this feature as : 

- we are not sure how it is supposed to work (especially the several invoices on one order feature) and i'm afraid everything we do could add new undesired behaviour changes
- there is still a workaround for the merchant : he can always modify invoices one by one manually

Is this ok for you @zuk3975 ?

zuk3975 commented:
> Is this ok for you @zuk3975 ?

yep, so Ill drop the "update all invoices feature"

zuk3975 committed:
remove ability to update invoices with AddCartRuleToOrderHandler and adjust tests

zuk3975 committed:
remove updating invoices from template and js

zuk3975 committed:
fix test wrong method used

matks reviewed (dismissed):
Code looks good to me 👍 
Re-running PrettyCI and Travis to see if issues are real issues or random glabberish from drunken CI

zuk3975 commented:
Removed the feature "Update all invoices". Reworked calculations using Number().

zuk3975 committed:
move specific price creation to private method

zuk3975 committed:
add Number DTO`s in adapter to help with calculations. Rework calculations in AddProductToOrderHandler to use Number from ps decimals

zuk3975 committed:
remove todo

zuk3975 committed:
implement calcualtions with Number in DeleteProductFromOrderHandler

zuk3975 committed:
remove checked todo

zuk3975 committed:
remove irrelevant code ant todo

zuk3975 committed:
implements Number calculations in UpdateProductInOrderHandler

zuk3975 committed:
applies private method for casting to Number

 apply cs fixer

zuk3975 committed:
get order object from abstract handler

zuk3975 commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -170,49 +146,40 @@ public function handle(AddProductToOrderCommand $command)
             $specificPrice->delete();
         }
 
-        $order = $order->refreshShippingCost();
```
todo - recheck this. Depends from configurations

zuk3975 committed:
fix Add behats for DeleteProductFromOrder

zuk3975 committed:
remove redundant var defining

zuk3975 committed:
add refreshShippingCost method which was removed previously

zuk3975 committed:
fix shipping price in AddProductToOrderHandler. Add some test cases

zuk3975 committed:
manual shipping cost calculation when free shipping

zuk3975 committed:
add todos for uncertain code

zuk3975 commented on src/Adapter/Order/CommandHandler/DeleteProductFromOrderHandler.php:
```diff
@@ -44,53 +46,70 @@
      */
     public function handle(DeleteProductFromOrderCommand $command)
     {
+        //@todo: what if its the last product, should order stay with zero prices or be cancelled?
```
@matks need help about this one.

zuk3975 commented on src/Adapter/Order/CommandHandler/AddProductToOrderHandler.php:
```diff
@@ -498,27 +440,163 @@ private function createNewInvoice(Order $order, Cart $cart, $isFreeShipping)
+     *  using  Order::refreshShippingCost() would end up shipping being $7
+     * @todo: If i do it manually, i don't update OrderCarrier shipping price.
+     *      Difference occurs between order shipping price and carrier shipping price
+     *      That might introduce problems. How To fix it?
```
@matks I have a problem with this (explained in TODO). If its unclear ill explain in chat. Could you help?

zuk3975 replied:
It would be easily solved if we apply shipping price to whole cart instead of one product. Or completely remove the option to create new invoice when adding product, or remove the free shipping option when adding product with new invoice. Idk all those features are very strange and doesn't comply with the system design

zuk3975 replied:
this was discussed in standup and chat. Decided to leave is at it was in legacy and dive deeper in later versions.

zuk3975 committed:
rollback AddProductToOrderHandler with Order::refreshFreeShipping

zuk3975 committed:
remove unused constructor arg

 cs fixer

zuk3975 committed:
adds todo

zuk3975 committed:
fix comparison in js which failed to display `newInvoice` block

zuk3975 committed:
rebuild assets

zuk3975 committed:
fix showing free shipping checkbox on addProduct with new invoice

zuk3975 committed:
remove adding cartrule twice

zuk3975 committed:
remove todo

Progi1984 commented on admin-dev/themes/new-theme/js/pages/order/view/order-product-renderer.js:
```diff
@@ -216,7 +216,7 @@ export default class OrderProductRenderer {
 
   toggleProductAddNewInvoiceInfo() {
-    if ($(OrderViewPageMap.productAddInvoiceSelect).val() === 0) {
+    if (parseInt($(OrderViewPageMap.productAddInvoiceSelect).val()) === 0) {
```
```suggestion
    if (parseInt($(OrderViewPageMap.productAddInvoiceSelect).val(), 10) === 0) {
```

zuk3975 replied:
is it necessary ? I get same results with or without it :thinking:

Progi1984 commented on src/Core/Domain/CartRule/ValueObject/PercentageDiscount.php:
```diff
@@ -33,6 +33,8 @@
  */
 class PercentageDiscount
 {
+    const MAX_PERCENTAGE = 100;
```
Why don't add a MIN_PERCENTAGE ?

zuk3975 replied:
what would it be? Min percentage should be just greater than zero, but not the zero itself. So would it be good to make MIN_PERCENTAGE = 0.000001 ?

Progi1984 commented on tests/Integration/Behaviour/Features/Context/Domain/OrderFeatureContext.php:
```diff
@@ -530,6 +662,89 @@ public function checkProductStockDifference(int $productDifference, string $fact
+    {
+        $configName = 'PS_ORDER_RECALCULATE_SHIPPING';
+
+        Configuration::set($configName, $status);
```
```suggestion
        Configuration::set('PS_ORDER_RECALCULATE_SHIPPING', $status);
```

zuk3975 committed:
remove redundant var

zuk3975 committed:
add radix to parseInt

zuk3975 committed:
re-apply changes from #17877

zuk3975 committed:
apply cs fixer

zuk3975 committed:
remove default precision value

zuk3975 commented:
The conflicts are too harsh. I don't know how to fix it anymore! :bomb:

matks commented:
@zuk3975 After deep analysis of this PR, I suggest the following plan to re-build it and deliver the improvements

## Scope summary

First, I summarized the large scope of this PR like this:

**Initial scope**

- fix `@todo` in **AddCartRuleToOrderHandler** ('move to private method' in couple places)

**Scope expansion**

- refactors **AddCartRuleToOrderHandler**, **AddProductToOrderHandler**, **DeleteProductFromOrderHandler**, and **UpdateProductInOrderHandler** handlers into private methods
- implement Number calculations instead of floats for handlers **AddCartRuleToOrderHandler**, **AddProductToOrderHandler**, **DeleteProductFromOrderHandler**, and **UpdateProductInOrderHandler**
- added some behat tests
- removed the ability to update invoices when adding discount to order _as requested by PM_
- fixed bug where "free shipping" checkbox wasn't displayed when adding product with new invoice

Please tell me if I was wrong on this scope.

## Roadmap to merge

I suggest to pick pieces of this PR and deliver them as new PRs, from new branches, one piece at a time. What I mean is in fact __looking at the diff and copy-pasting pieces of code into new branches created from 1.7.7.X__.

I suggest doing this in this order:
1. Pick the code necessary to remove the ability to update invoices when adding discount to order ; pick related behat tests if there is one ; create a PR for it and only for this. Pick **the smallest possible amount of code** following a TDD fashion. No refactoring in this PR, no conversion from float to numbers.
2. Pick the code necessary to fix the bug where "free shipping" checkbox was not displayed after adding product with new invoice. Create a PR with this piece of code. Again pick **the smallest possible amount of code**. I would like that **this is impossible to remove any of the committed lines without breaking the bugfix.**

These 2 items are the necessary items for 177x. Following items can be done 178 (develop). We'll see when we reach this step how far we are in 177 release process.

3. Pick the code to refactor AddCartRuleToOrderHandler using more functions (split big functions into small ones ; pick related behat tests if there are some. Use the behat tests to validate we break nothing. Create a PR for that. No other refactoring in this PR, no conversion from float to numbers.
4. Do the same for AddProductToOrderHandler, DeleteProductFromOrderHandler, and UpdateProductInOrderHandler, one PR for each.
5. Finally pick the code to refactor AddCartRuleToOrderHandler using Numbers. At this point I expect all behat tests to have been merged before, allowing them to make sure we break nothing. Create a PR for that.
6. Do the same for AddProductToOrderHandler, DeleteProductFromOrderHandler, and UpdateProductInOrderHandler, one PR for each.

@zuk3975 This is a long process and you might be too busy to do it. I will share the workload if necessary.

zuk3975 commented:
> I suggest doing this in this order:
> 
> 1. Pick the code necessary to remove the ability to update invoices when adding discount to order ...
> Pick the code necessary to fix the bug where "free shipping" checkbox was not displayed after adding product with new invoice....

1. #18524

zuk3975 commented:
Created a reminder issue for 1.7.8 #18545. Closing this pr.
